sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/V2TableUtil.scala

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog
+
+import java.util.Locale
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.expressions.AttributeReference
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.util.ArrayImplicits._
+
+private[sql] object V2TableUtil extends SQLConfHelper {
+
+  def validateCapturedColumns(table: Table, relation: DataSourceV2Relation): Seq[String] = {
+    validateCapturedColumns(table, relation.table.columns.toImmutableArraySeq)
+  }
+
+  def validateCapturedColumns(table: Table, originCols: Seq[Column]): Seq[String] = {
+    val errors = mutable.ArrayBuffer[String]()
+
+    val colsByNormalizedName = indexByNormalizedName(table.columns.toImmutableArraySeq)
+    val originColsByNormalizedName = indexByNormalizedName(originCols)
+
+    originColsByNormalizedName.foreach { case (normalizedName, originCol) =>
+      colsByNormalizedName.get(normalizedName) match {
+        case Some(col) =>
+          if (originCol.dataType != col.dataType || originCol.nullable != col.nullable) {
+            val oldType = formatType(originCol.dataType, originCol.nullable)
+            val newType = formatType(col.dataType, col.nullable)
+            errors += s"`${originCol.name}` type has changed from $oldType to $newType"
+          }
+        case None =>
+          errors += s"${formatColumn(originCol)} is missing"
```

Comment:
should it be `s"${formatColumn(originCol)} has been deleted"` to be consistent with the other error message?